lightly_studio/src/lightly_studio/dataset/perception_encoder_embedding_generator.py

```diff
@@ -49,7 +51,78 @@ def __getitem__(self, idx: int) -> torch.Tensor:
             return self.preprocess(image)
 
 
-class PerceptionEncoderEmbeddingGenerator(ImageEmbeddingGenerator):
+class _VideoFileDataset(Dataset[torch.Tensor]):
+    """Dataset wrapping video file paths and a preprocess function.
+
+    Used for efficient batched video loading and preprocessing
+    """
+
+    def __init__(
+        self,
+        filepaths: list[str],
+        preprocess: Callable[[Image.Image], torch.Tensor],
+    ) -> None:
+        self.filepaths = filepaths
+        self.preprocess = preprocess
+
+    def __len__(self) -> int:
+        return len(self.filepaths)
+
+    def __getitem__(self, idx: int) -> torch.Tensor:
+        """Return tensor [N C H W] for idx-th video.
+
+        As in the original paper we subsample N frames from a video and stack them to a tensor.
+        As in the paper, we use a default of 8 frames per video (VIDEO_FRAMES_PER_SAMPLE).
+        Note: the video length in the paper was 16.7 +/- 9.8 sec, hence for longer videos we might
+        consider alternative models or more frames.
+        """
+        video_path = self.filepaths[idx]
+        frames = self._load_frames(video_path)
+        if not frames:
+            raise ValueError(f"Unable to read frames from video '{video_path}'.")
+
+        processed_frames = [self.preprocess(frame) for frame in frames]
+        return torch.stack(processed_frames)
+
+    def _load_frames(self, video_path: str) -> list[Image.Image]:
+        """Sample uniformly spaced frames and return them as PIL images.
+
+        Using seek for sampling is fast, however it may yield slightly different results on
+        different OS (known issue: MacOS vs Linux).
+
+        Alternative option is to decode frame-by-frame to be OS independent,
+        however this comes with performance drop.
+        """
+        fs, fs_path = fsspec.core.url_to_fs(url=video_path)
+        with fs.open(path=fs_path, mode="rb") as video_file:  # noqa: SIM117
+            with container.open(file=video_file) as video_container:
```

Comment:
Optional: I looked up `SIM117`. It should be possible to simplify and merge the `with` statements into one:
```py
        with fs.open(path=fs_path, mode="rb") as video_file \
            container.open(file=video_file) as video_container:
```